scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py

```diff
@@ -90,36 +107,51 @@ def _b_orthonormalize(B, blockVectorV, blockVectorBV=None,
                         f"{e}\n",
                         UserWarning, stacklevel=3
                     )
-                    return None, None, None, normalization
+                    return None, None, None
             if blockVectorBV.shape != blockVectorV.shape:
                 raise ValueError(
                     f"The shape {blockVectorV.shape} "
                     f"of the orthogonalized matrix not preserved\n"
                     f"and changed to {blockVectorBV.shape} "
                     f"after multiplying by the secondary matrix.\n"
                 )
-        else:
-            blockVectorBV = blockVectorV  # Shared data!!!
-    else:
-        blockVectorBV = blockVectorBV / normalization
+
     VBV = blockVectorV.T.conj() @ blockVectorBV
     try:
         # VBV is a Cholesky factor from now on...
         VBV = cholesky(VBV, overwrite_a=True)
         VBV = inv(VBV, overwrite_a=True)
-        blockVectorV = blockVectorV @ VBV
-        # blockVectorV = (cho_solve((VBV.T, True), blockVectorV.T)).T
+        try:
+            np.matmul(blockVectorV, VBV, out=blockVectorV)
+        except Exception:
+            if verbosityLevel:
+                warnings.warn(
+                    f"The blockVectorX dtype {blockVectorV.dtype} "
+                    f"does not match {VBV.dtype} or is a slice "
+                    f"and needs to be changed preventing in-place.",
+                    UserWarning, stacklevel=3
+                )
+            blockVectorV = blockVectorV @ VBV
         if B is not None:
-            blockVectorBV = blockVectorBV @ VBV
-            # blockVectorBV = (cho_solve((VBV.T, True), blockVectorBV.T)).T
-        return blockVectorV, blockVectorBV, VBV, normalization
+            try:
+                np.matmul(blockVectorBV, VBV, out=blockVectorBV)
+            except Exception:
+                if verbosityLevel:
+                    warnings.warn(
+                        f"The blockVectorX dtype {blockVectorBV.dtype} "
+                        f"does not match {VBV.dtype} "
+                        f"and needs to be changed preventing in-place.",
+                        UserWarning, stacklevel=3
+                    )
+                blockVectorBV = blockVectorBV @ VBV
+        return blockVectorV, blockVectorBV, VBV
     except LinAlgError:
         if verbosityLevel:
             warnings.warn(
                 "Cholesky has failed.",
                 UserWarning, stacklevel=3
             )
-        return None, None, None, normalization
+        return None, None, None
```

Comment:
It's not obvious to me from the global diff (didn't go through 75 commits individually), how the change in the signature of `_b_orthonormalize` is related to performance considerations or increasing the amount of in-place operations.

Reply:
This is an actual change in the algorithm, not just in the signature. The vector `normalization` is no longer computed and all lines of the previous code that perform operations with it are being deleted, which evidently increases performance.

Reply:
@h-vetinari - resolved?

Reply:
Yes, fine.